### PR TITLE
speed up matching_methods for dispatch tuples

### PR DESCRIPTION
**Pull request by JeffBezanson**
`jl_matching_methods` is one of the inner loops in type inference. When the argument is `isdispatchtuple` (by far the common case) we can use a faster query since at most one method matches. This gives a very modest speedup of around 4-5%, but it affects almost everything and has basically no cost so it's nice anyway.

Time to first plot before:
```
 23.422304 seconds (46.91 M allocations: 2.295 GiB, 5.94% gc time)
```
after:
```
 21.891047 seconds (46.19 M allocations: 2.263 GiB, 6.56% gc time)
```

### CODE REVIEW

**Review by vtjnash — CHANGES_REQUESTED** [review]
I’ve wanted this many times too. But unfortunately since it doesn’t set the world-limit output fields, it cannot be used currently.

**Review by vtjnash — COMMENTED** [review]
(fwiw, even if that wasn’t the case, I think this would may still be the wrong place to put this logic, as all intersection visitors would probably want this optimization)

**Comment by JeffBezanson** [comment]
> But unfortunately since it doesn’t set the world-limit output fields, it cannot be used currently.

Ah, right. That seems easy to fix though --- can we just intersect the range with the values in the typemap entry that gets looked up?


> as all intersection visitors would probably want this optimization

Skipping the overhead of the intersection visitor is part of the optimization though, e.g. setting up the closure environment.

File: src/gf.c [hunk 1]
```diff
@@ -1800,6 +1800,19 @@ JL_DLLEXPORT jl_value_t *jl_matching_methods(jl_tupletype_t *types, int lim, int
     jl_methtable_t *mt = jl_method_table_for(unw);
     if ((jl_value_t*)mt == jl_nothing)
         return jl_false; // indeterminate - ml_matches can't deal with this case
+    if (jl_is_dispatch_tupletype((jl_value_t*)types)) {
+        JL_LOCK(&mt->writelock);
+        jl_method_instance_t *sf = jl_mt_assoc_by_type(mt, types, 1, world);
```
Comment (vtjnash) [inline src/gf.c]:
```suggestion
        jl_method_instance_t *sf = jl_mt_assoc_by_type(mt, types, 0, world);
```

**Comment by vtjnash** [comment]
> can we just intersect the range with the values in the typemap entry that gets looked up?

no: it's the range of the intersection, not the lookup.

**Comment by vtjnash** [comment]
Replaced by https://github.com/JuliaLang/julia/pull/34745

**Pull request closed without merging**